### PR TITLE
controller: add s3 backup

### PR DESCRIPTION
**Pull request by fanminshi**

### CODE REVIEW

**Comment by fanminshi** [comment]
cc/ @hongchaodeng @hasbro17

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	"crypto/tls"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/Sirupsen/logrus"
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: `github.com/coreos/etcd-operator` imports should be grouped together and be separate from non etcd-operator related imports.

Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: good catch. easy to miss those

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+		clusterName: clusterName,
+		s3bucket:    s3.S3Bucket,
+		prefix:      ToS3Prefix(s3.Prefix, b.namespace, clusterName),
+		awsSecret:   s3.AWSSecret,
+		kubecli:     b.kubecli,
+	}
+	return s3b.saveSnap()
+}
+
+const S3V1 = "v1"
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: Move all your consts and types to the top of the file. So it should be structured like:
```
const (
...
)

type ...

func ...

func ...
```

Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: done!

File: pkg/etcd-backup-operator/controller/s3backup.go (hunk 1 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: done!

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	if len(pods) == 0 {
+		msg := "no running etcd pods found"
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	member, rev := getMemberWithMaxRev(pods, nil)
+	if member == nil {
+		logrus.Warning("no reachable member")
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: To keep it consistent as above:
```
msg := "no reachable member"
logrus.Warning(msg)
return fmt.Errorf(msg)
```

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		return fmt.Errorf("no reachable member")
+	}
+
+	log.Printf("saving backup for cluster (%s)", b.clusterName)
+	if err := b.writeSnap(member, rev); err != nil {
+		err = fmt.Errorf("write snapshot failed: %v", err)
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: `return fmt.Errorf("write snapshot failed: %v", err)`

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	defer rc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to receive snapshot (%v)", err)
+	}
+	return b.saveToS3(resp.Version, rev, rc)
+}
+
+// NewS3Uploader returns a Uploader that is configured with proper aws config and credentials derived from Kubernetes secret.
+func NewS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: This function does not need to be exported right? So it should be lowercase `newS3Uploader`.

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	member, rev := getMemberWithMaxRev(pods, nil)
+	if member == nil {
+		logrus.Warning("no reachable member")
+		return fmt.Errorf("no reachable member")
+	}
+
+	log.Printf("saving backup for cluster (%s)", b.clusterName)
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/s3backup.go]: Use logrus

File: pkg/etcd-backup-operator/controller/sync.go [hunk 7]
```diff
@@ -48,11 +48,19 @@ func (b *Backup) processItem(key string) error {
...
 	// Remove fmt.Println(eb) when implementing handle backup logic.
-	fmt.Println(eb)
-	// TODO: handle backup
+	b.handleBackup(&eb.Spec)
 	return nil
 }
 
+func (b *Backup) handleBackup(spec *api.EtcdBackupSpec) error {
+	switch spec.StorageType {
+	case "s3":
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/sync.go]: Use const?

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("setup AWS config failed: write config file failed: %v", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, configFile)
+	}
+
+	return options, nil
+}
+
+func NewFromSessionOpt(bucket, prefix string, so session.Options) (*s3manager.Uploader, error) {
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: This function should also be unexported and lowercased.

File: pkg/etcd-backup-operator/controller/s3backup.go (hunk 5 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: yeah, you are right.

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+	clusterName string
+
+	s3bucket  string
+	prefix    string
+	awsSecret string
+
+	kubecli kubernetes.Interface
+}
+
+func (b *s3Backup) saveSnap() error {
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: We're already using the receiver `(b *Backup)` above.
Plus the naming convention for method receivers is that generally it should be the first one or two letters of its type. So can you rename them to `(s *s3Backup)`

Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: agreed.

File: pkg/etcd-backup-operator/controller/s3backup.go (hunk 8 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: fixed

File: pkg/etcd-backup-operator/controller/sync.go (hunk 7 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/sync.go]: fixed

File: pkg/etcd-backup-operator/controller/s3backup.go (hunk 6 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: fixed

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return newFromSessionOpt(bucket, prefix, *so)
+}
+
+func newAWSConfig(secret *v1.Secret, prefix string) (*session.Options, error) {
+	dir := filepath.Join(operatorRoot, "aws", prefix)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, err
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: Probably wrap most errors like you've done on line 152 for better context:
`return nil, fmt.Errorf("setup AWS config failed: mkdir (%v) failed: %v", dir, err)`
Can you also double check anywhere else for the same.

**Comment by fanminshi** [comment]
all fixed. PTAL cc/ @hasbro17

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+	"github.com/coreos/etcd-operator/pkg/util/constants"
+	"github.com/coreos/etcd-operator/pkg/util/etcdutil"
+	"github.com/coreos/etcd-operator/pkg/util/k8sutil"
+	"github.com/coreos/etcd/clientv3"
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: Missed this one :P

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
+package controller
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/s3backup.go]: license header

File: pkg/etcd-backup-operator/controller/sync.go [hunk 13]
```diff
@@ -14,6 +14,7 @@ const (
 	//
 	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
 	maxRetries = 15
+	s3         = "s3"
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/sync.go]: there is already a StorageType for s3 defined

Comment (hongchaodeng) [reply pkg/etcd-backup-operator/controller/sync.go]: BackupStorageTypeS3               = "S3"

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+const (
+	S3V1                 = "v1"
+	operatorRoot         = "/var/tmp/etcd-backup-operator"
+	backupFilenameSuffix = "etcd.backup"
+)
+
+type s3Backup struct {
+	namespace   string
+	name        string
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/s3backup.go]: What is this 'name' for?

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+			maxRev = resp.Header.Revision
+			member = m
+		}
+	}
+	return member, maxRev
+}
+
+// ToS3Prefix concatenates s3Prefix, S3V1, namespace, clusterName to a single s3 prefix.
+// the concatenated prefix determines the location of S3 backup files.
+func ToS3Prefix(s3Prefix, namespace, clusterName string) string {
```
Comment (hasbro17) [inline pkg/etcd-backup-operator/controller/s3backup.go]: Make this unexported

File: pkg/etcd-backup-operator/controller/s3backup.go (hunk 14 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: not used anywhere. will delete.

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+func newS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
+	so, err := newAWSConfig(secret, prefix)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup aws config: (%v)", err)
+	}
+	return newFromSessionOpt(bucket, prefix, *so)
+}
+
+func newAWSConfig(secret *v1.Secret, prefix string) (*session.Options, error) {
+	dir := filepath.Join(operatorRoot, "aws", prefix)
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/s3backup.go]: If operatorRoot is only used for storing temporary credentials, I would suggest we just store it under "/tmp" and make tempdir there.

Comment (hongchaodeng) [reply pkg/etcd-backup-operator/controller/s3backup.go]: Also this codepath should be refactored and reused.
This is unnecessary duplicate.

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to write config file: (%v)", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, configFile)
+	}
+
+	return options, nil
+}
+
+func newFromSessionOpt(bucket, prefix string, so session.Options) (*s3manager.Uploader, error) {
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/s3backup.go]: this is newS3Uploader

File: pkg/etcd-backup-operator/controller/s3backup.go [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+	sess, err := session.NewSessionWithOptions(so)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new AWS session: (%v)", err)
+	}
+	return s3manager.NewUploader(sess), nil
+}
+
+func getMemberWithMaxRev(pods []*v1.Pod, tc *tls.Config) (*etcdutil.Member, int64) {
```
Comment (hongchaodeng) [inline pkg/etcd-backup-operator/controller/s3backup.go]: This is duplicate code.
Please refactor and reuse.
Same for every other potential duplicate code.

File: pkg/etcd-backup-operator/controller/sync.go (hunk 13 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/sync.go]: do you mean here? https://github.com/coreos/etcd-operator/blob/6a8e08ffc11cfc4dcd4f178ad1664cde16971bcb/pkg/apis/etcd/v1beta2/backup.go#L24

File: pkg/etcd-backup-operator/controller/s3backup.go (hunk 18 above)
Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: @hongchaodeng i think this one is fine since the other `getMemberWithMaxRev` will be deleted.

Comment (hongchaodeng) [reply pkg/etcd-backup-operator/controller/s3backup.go]: I see what you are trying to do...

Please refactor and reuse the codebase in pkg/backup/ in both places.

Comment (fanminshi) [reply pkg/etcd-backup-operator/controller/s3backup.go]: kk

**Comment by fanminshi** [comment]
@hongchaodeng @hasbro17 do we want to support TLS for talking to etcd cluster in this pr?

**Comment by hongchaodeng** [comment]
@fanminshi 
Let's deal with that later to make smaller PR at a time :)

**Comment by fanminshi** [comment]
@hongchaodeng sounds good.

**Comment by fanminshi** [comment]
@hongchaodeng refactor the code. ptal

File: pkg/util/backuputil/backuptutil.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/coreos/etcd-operator/pkg/util/constants"
+	"github.com/coreos/etcd-operator/pkg/util/etcdutil"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/coreos/etcd/clientv3"
+	"k8s.io/api/core/v1"
+)
+
+func NewAWSConfig(secret *v1.Secret, tmpDir string) (*session.Options, error) {
```
Comment (xiang90) [inline pkg/util/backuputil/backuptutil.go]: please document public functions.

Comment (xiang90) [reply pkg/util/backuputil/backuptutil.go]: why this lives inside backuputil? this has nothing to do with backup specifically.

File: pkg/util/backuputil/backuptutil.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+		err := ioutil.WriteFile(configFile, config, 0600)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write config file: (%v)", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, configFile)
+	}
+	return options, nil
+}
+
+func GetMemberWithMaxRev(pods []*v1.Pod, tc *tls.Config) (*etcdutil.Member, int64) {
```
Comment (xiang90) [inline pkg/util/backuputil/backuptutil.go]: document public funcs. and this func probably should not live inside backuputil. it is not backup related.

File: pkg/util/backuputil/backuptutil.go (hunk 19 above)
Comment (fanminshi) [reply pkg/util/backuputil/backuptutil.go]: maybe awsutil?

File: pkg/util/backuputil/backuptutil.go (hunk 20 above)
Comment (fanminshi) [reply pkg/util/backuputil/backuptutil.go]: let me think about a better place to place it

**Pull request closed without merging**